packages/lib/utils/chunkGenerator.ts

```diff
@@ -0,0 +1,50 @@
+import * as fs from 'fs';
+
+function readBytes(fd: number, sharedBuffer: Buffer) {
+	return new Promise<void>((resolve, reject) => {
+		fs.read(fd, sharedBuffer, 0, sharedBuffer.length, null, (error) => {
+			if (error) {
+				return reject(error);
+			}
+			resolve();
+		});
+	});
+}
+
+async function* generateChunks(filePath: string, size: number) {
+	const sharedBuffer = Buffer.alloc(size);
+	let stats;
+	let fd;
+	let bytesRead = 0;
+	let end = size;
+
+
+	try {
+		stats = fs.statSync(filePath);
+		fd = fs.openSync(filePath, 'r');
+	} catch (_) {
+		throw new Error('Failed to open file');
+	}
+
+	const chunksNo = Math.ceil(stats.size / size);
+	for (let i = 0; i < chunksNo; i++) {
+		try {
+			await readBytes(fd, sharedBuffer);
+		} catch (_) {
+			throw new Error('Failed to read bytes from file');
+		}
+
+		bytesRead = (i + 1) * size;
+		if (bytesRead > stats.size) {
+			end = size - (bytesRead - stats.size);
+		}
+
+		yield {
+			index: i,
+			chunksNo: chunksNo,
+			chunk: sharedBuffer.slice(0, end),
+		};
+	}
+}
+
+export default generateChunks;
```

Comment:
I'm struggling to be ok with code like this. Is that generated by ChatGPT or something? It has "yield", weird promisified functions, fs sync and async code, errors are silenced, variables all declared on top as if it was C code, etc.

Again I have zero confidence that this pull request works as expected. I don't want to merge and have to revert in emergency when it starts breaking people's Dropbox sync targets.

Reply:
it's not gpt generated, but it's from a research of approaching this, I used the promisified approach like this just for the readability and I can make it an inline code but didn't prefer it! 

if you have any guides for me regarding my approach, I'm open to it! 

ps: I don't use gpt, last time I used it was like 5 months ago, it produce shitty answers and code... yes, maybe i write shitty code and I take snippets from other places but calling it gpt generated is kinda frustrating.

Reply:
Ok so a good starting point then would be to read our coding style guide. I could go over the details but most of it is covered in our coding style - use of sync functions, variable declaration, how to name files, etc. Some of it is just styling preference, but others are more to do with code correctness and it means you might learn a useful thing or two which will help you write better code.

Reply:
Thanks Laurent!